ext/random/randomizer.c

```diff
@@ -88,6 +89,101 @@ PHP_METHOD(Random_Randomizer, __construct)
 }
 /* }}} */
 
+/* {{{ Generate a float in [0, 1) */
+PHP_METHOD(Random_Randomizer, nextFloat)
+{
+	php_random_randomizer *randomizer = Z_RANDOM_RANDOMIZER_P(ZEND_THIS);
+	uint64_t result;
+	size_t total_size;
+
+	ZEND_PARSE_PARAMETERS_NONE();
+
+	result = 0;
+	total_size = 0;
+	do {
+		uint64_t r = randomizer->algo->generate(randomizer->status);
+		result = result | (r << (total_size * 8));
+		total_size += randomizer->status->last_generated_size;
+		if (EG(exception)) {
+			RETURN_THROWS();
+		}
+	} while (total_size < sizeof(uint64_t));
+
+	/* A double has 53 bits of precision, thus we must not
+	 * use the full 64 bits of the uint64_t, because we would
+	 * introduce a bias / rounding error.
+	 */
+#if DBL_MANT_DIG != 53
+# error "Random_Randomizer::nextFloat(): Requires DBL_MANT_DIG == 53 to work."
+#endif
+	const double step_size = 1.0 / (1ULL << 53);
+
+	/* Use the upper 53 bits, because some engine's lower bits
+	 * are of lower quality.
+	 */
+	result = (result >> 11);
+
+	RETURN_DOUBLE(step_size * result);
+}
+/* }}} */
+
+/* {{{ Generates a random float within a configurable interval.
+ *
+ * This method uses the γ-section algorithm by Frédéric Goualard.
+ */
+PHP_METHOD(Random_Randomizer, getFloat)
+{
+	php_random_randomizer *randomizer = Z_RANDOM_RANDOMIZER_P(ZEND_THIS);
+	double min, max;
+	zend_object *bounds = NULL;
+	zend_string *bounds_name = zend_string_init("ClosedOpen", strlen("ClosedOpen"), 1);
+
+	ZEND_PARSE_PARAMETERS_START(2, 3)
+		Z_PARAM_DOUBLE(min)
+		Z_PARAM_DOUBLE(max)
+		Z_PARAM_OPTIONAL
+		Z_PARAM_OBJ_OF_CLASS(bounds, random_ce_Random_IntervalBoundary);
+	ZEND_PARSE_PARAMETERS_END();
+
+	if (bounds) {
+		zval *case_name = zend_enum_fetch_case_name(bounds);
+		bounds_name = Z_STR_P(case_name);
+	}
+	
+	if (zend_string_equals_literal(bounds_name, "ClosedOpen")) {
```

Comment:
Do we really need to compare the names here, or is there maybe a more performant option? cc @iluuu1994 

If we need to compare the names, it might make sense to have a switch dispatching on the first character ('C' or 'O') in advance. Or possible even two switches, assuming the enum won't change, in which case we wouldn't even need to compare strings.

Reply:
Currently no. Maybe we can fetch constants by offset like we do for properties (`OBJ_PROP_NUM()` macro), and then do an identity check?

Reply:
@TimWolla I think you may be able to access the constants table by offset (since it's a HashMap which are ordered), assuming internal enums initialize all constants in Minit which I think they do. In that case you could then compare the enum object by pointer.

Reply:
@iluuu1994 I'm not sure how I would do that; in any case it sounds more fragile than my current solution of building a simple "hash", because it would be sensitive to reordering (instead of just renaming / extending)?

Reply:
> since we have tests for these cases

The tests can't really distinguish between the cases, because the difference only matter (literal!) edge cases and by the nature of randomness you can't reliably hit those. So in this case at least we need to rely on careful review and making the code resilient against *accidental* changes.

> Anyway, we haven't established a standard here

Yeah, it's the first native enum. I tend to hit those “first times” with my RFCs, `#[SensitiveParameter]` was the first parameter attribute and also required some general fixes for memory leaks in parameter attributes …

Reply:
> my current solution of building a simple "hash"

Interesting solution! :)

> Maybe we could make the stubs generator generate macros for the case offsets so this becomes less fragile.

@kocsismate, thoughts?